### PR TITLE
feat: add Recursive tree traversal techniques

### PR DESCRIPTION
**Pull request by Lazeeez**
#### Description of Change

All Tree traversal techniques (inorder, postorder, preorder) with recursive functions.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: Added tree traversal implementations with recursive functions.

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1865"></a>

### CODE REVIEW

**Comment by Lazeeez** [comment]
@Panquesito7 can you check the PR now, thanks.

**Review by Qannaf — APPROVED** [review]

File: others/recursive_tree_traversal.cpp [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+    postorder(root->left);   // Traverse the left subtree
+    postorder(root->right);  // Traverse the right subtree
+    BT::postorder_result.push_back(
+        root->data);  // Display the data part of the root (or current node)
+
+    return postorder_result;
+}
+
+}  // namespace recursive_tree_traversals
+
```
Comment (Panquesito7) [inline others/recursive_tree_traversal.cpp:184]:
```suggestion
```

File: others/recursive_tree_traversal.cpp [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+    std::cout << "3rd test-case" << std::endl;
+    test3();  // run 3rd test-case
+}
+
+/*******************************************************************************
+ * @brief Main function
+ * @returns 0 on exit
+ *******************************************************************************/
+int main() {
+    tests();
```
Comment (Panquesito7) [inline others/recursive_tree_traversal.cpp:375]:
```suggestion
    tests();  // run self-test implementations
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 😄

**Comment by Lazeeez** [comment]
@Panquesito7 changes done.

**Comment by Panquesito7** [comment]
> @Panquesito7 changes done.

Hey there! Hmm, I don't see any new commits 👀 I think you forgot to push or something, not sure. 😅

**Comment by Lazeeez** [comment]
@Panquesito7 check now, had to manually do all the changes.

**Review by Panquesito7 — DISMISSED** [review]
LGTM 🚀 Thank you for your contribution! 😄👍

File: others/recursive_tree_traversal.cpp [hunk 3]
```diff
@@ -0,0 +1,377 @@
+/******************************************************************************
```
Comment (mishraabhinn) [inline others/recursive_tree_traversal.cpp:1]:
```suggestion
/**

File: others/recursive_tree_traversal.cpp [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+ *
+ * @author [Lajat Manekar](https://github.com/Lazeeez)
+ *
+ ******************************************************************************/
+
+#include <cassert>   /// for assert
+#include <iostream>  /// for I/O operations
+#include <vector>    /// for vector
+
+/******************************************************************************
```
Comment (mishraabhinn) [inline others/recursive_tree_traversal.cpp:58]:
```suggestion
/**
```

File: others/recursive_tree_traversal.cpp [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+ * @namespace others
+ * @brief Other algorithms
+ *******************************************************************************/
```
Comment (mishraabhinn) [inline others/recursive_tree_traversal.cpp:61]:
```suggestion
 */
```

File: others/recursive_tree_traversal.cpp [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+namespace others {
+
+/******************************************************************************
+ * @namespace interpolation_search
+ * @brief Functions for the Recursive version of Inorder, Preorder, and
+ * Postorder [Traversal of the
+ * Tree](https://en.wikipedia.org/wiki/Tree_traversal) algorithm implementation
+ *******************************************************************************/
```
Comment (mishraabhinn) [inline others/recursive_tree_traversal.cpp:69]:
```suggestion
/**
 * @namespace interpolation_search
 * @brief Functions for the Recursive version of Inorder, Preorder, and
 * Postorder [Traversal of the
 * Tree](https://en.wikipedia.org/wiki/Tree_traversal) algorithm implementation
 */
```

File: others/recursive_tree_traversal.cpp [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+namespace recursive_tree_traversals {
+
+/******************************************************************************
+ * @brief The structure to hold Nodes of the tree.
+ * @param data Value that will be stored in the node.
+ * @param left follow up left subtree.
+ * @param right follow up right subtree.
+ *******************************************************************************/
```
Comment (mishraabhinn) [inline others/recursive_tree_traversal.cpp:77]:
```suggestion
/**
 * @brief The structure to hold Nodes of the tree.
 * @param data Value that will be stored in the node.
 * @param left follow up left subtree.
 * @param right follow up right subtree.
 */
```

File: others/recursive_tree_traversal.cpp [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+struct Node {
+    uint64_t data = 0;     ///< The value/key of the node.
+    struct Node *left{};   ///< struct pointer to left subtree.
+    struct Node *right{};  ///< struct pointer to right subtree.
+};
+/******************************************************************************
+ * @brief BT used to make the entire structure of the binary tree and the
+ * functions associated with the binary tree
+ *******************************************************************************/
```
Comment (mishraabhinn) [inline others/recursive_tree_traversal.cpp:86]:
```suggestion
/**
 * @brief BT used to make the entire structure of the binary tree and the
 * functions associated with the binary tree
 */
```

File: others/recursive_tree_traversal.cpp [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+        Node *);  // function that takes root of the tree as an argument and
+                  // returns its postorder traversal.
+};
+
+/******************************************************************************
+ * @brief will allocate the memory for a node and, along the data and return the
+ * node.
+ * @param data value that a particular node will contain.
+ * @return pointer to the newly created node with assigned data.
+ *******************************************************************************/
```
Comment (mishraabhinn) [inline others/recursive_tree_traversal.cpp:115]:
```suggestion
/**
 * @brief will allocate the memory for a node and, along the data and return the
 * node.
 * @param data value that a particular node will contain.
 * @return pointer to the newly created node with assigned data.
 */
```

File: others/recursive_tree_traversal.cpp [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+    return node;
+}
+
+/******************************************************************************
+ * @brief inorder() function that will perform the inorder traversal
+ * recursively, and return the resultant vector that contain the inorder
+ * traversal of a tree.
+ * @param root head/root node of a tree
+ * @return result that is containing the inorder traversal of a tree
+ *******************************************************************************/
```
Comment (mishraabhinn) [inline others/recursive_tree_traversal.cpp:129]:
```suggestion
/**
 * @brief inorder() function that will perform the inorder traversal
 * recursively, and return the resultant vector that contain the inorder
 * traversal of a tree.
 * @param root head/root node of a tree
 * @return result that is containing the inorder traversal of a tree
 */
```

File: others/recursive_tree_traversal.cpp [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+    return inorder_result;
+}
+
+/******************************************************************************
+ * @brief preorder function that will perform the preorder traversal
+ * recursively, and return the resultant vector that contain the preorder
+ * traversal of a tree.
+ * @param root head/root node of a tree
+ * @return result that is containing the preorder traversal of a tree
+ *******************************************************************************/
```
Comment (mishraabhinn) [inline others/recursive_tree_traversal.cpp:149]:
```suggestion
/**
 * @brief preorder function that will perform the preorder traversal
 * recursively, and return the resultant vector that contain the preorder
 * traversal of a tree.
 * @param root head/root node of a tree
 * @return result that is containing the preorder traversal of a tree
 */
```

File: others/recursive_tree_traversal.cpp [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+    return preorder_result;
+}
+
+/******************************************************************************
+ * @brief postorder function that will perform the postorder traversal
+ * recursively, and return the result vector that contain the postorder
+ * traversal of a tree.
+ * @param root head/root node of a tree
+ * @return result that is containing the postorder traversal of a tree
+ *******************************************************************************/
```
Comment (mishraabhinn) [inline others/recursive_tree_traversal.cpp:169]:
```suggestion
/**
 * @brief postorder function that will perform the postorder traversal
 * recursively, and return the result vector that contain the postorder
 * traversal of a tree.
 * @param root head/root node of a tree
 * @return result that is containing the postorder traversal of a tree
 */
```

File: others/recursive_tree_traversal.cpp [hunk 13]
```diff
@@ -0,0 +1,377 @@
...
+}
+
+}  // namespace recursive_tree_traversals
+
+}  // namespace others
+
+/*******************************************************************************
+ * @brief 1st test-case
+ * @returns void
+ *******************************************************************************/
```
Comment (mishraabhinn) [inline others/recursive_tree_traversal.cpp:190]:
```suggestion
/**
 * @brief 1st test-case
 * @returns void
 */
```

File: others/recursive_tree_traversal.cpp [hunk 14]
```diff
@@ -0,0 +1,377 @@
...
+    }
+    std::cout << "Passed!" << std::endl;
+
+    std::cout << std::endl;
+}
+
+/*******************************************************************************
+ * @brief 2nd test-case
+ * @returns void
+ *******************************************************************************/
```
Comment (mishraabhinn) [inline others/recursive_tree_traversal.cpp:248]:
```suggestion
/**
 * @brief 2nd test-case
 * @returns void
 */
```

File: others/recursive_tree_traversal.cpp [hunk 15]
```diff
@@ -0,0 +1,377 @@
...
+    std::cout << "2nd test-case" << std::endl;
+    test2();  // run 2nd test-case
+    std::cout << "3rd test-case" << std::endl;
+    test3();  // run 3rd test-case
+}
+
+/*******************************************************************************
+ * @brief Main function
+ * @returns 0 on exit
+ *******************************************************************************/
```
Comment (mishraabhinn) [inline others/recursive_tree_traversal.cpp:373]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
```

File: others/recursive_tree_traversal.cpp [hunk 16]
```diff
@@ -0,0 +1,377 @@
...
+    }
+    std::cout << "Passed!" << std::endl;
+
+    std::cout << std::endl;
+}
+
+/*******************************************************************************
+ * @brief 3rd test-case
+ * @returns void
+ *******************************************************************************/
```
Comment (mishraabhinn) [inline others/recursive_tree_traversal.cpp:305]:
```suggestion
/**
 * @brief 3rd test-case
 * @returns void
 */
```

**Review by mishraabhinn — CHANGES_REQUESTED** [review]
I feel this makes it cleaner to read. If you like this @Lazeeez @Panquesito7  we can change it as suggested thanks.

**Comment by Lazeeez** [comment]
@mishraabhinn changes made, will keep it in mind next time, thanks.

**Review by mishraabhinn — APPROVED** [review]

**Review by Panquesito7 — APPROVED** [review]

**Pull request merged**